From cc7e73bf8a0fa03409125df7a906c1bbcbf533f9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Niklas=20Ekstr=C3=B6m?= <mail@niklasekstrom.nu>
Date: Mon, 3 Jun 2024 18:02:58 +0200
Subject: [PATCH 13/13] Add more uniform naming of libremote files


diff --git a/.gitignore b/.gitignore
index e09d7c1..6d5e024 100644
--- a/.gitignore
+++ b/.gitignore
@@ -11,6 +11,7 @@ Software/bin_amiga
 Software/bin_pi
 Software/bpls2gif/build
 Software/libremote/stubs.h
-Software/libremote/proto_lib.h
 Software/libremote/example.library
+Software/libremote/bsdsocket.library
+Software/libremote/proto_example.h
 Software/libremote/ExampleClient
diff --git a/Software/libremote/bsl.py b/Software/libremote/bsdsocket.py
similarity index 100%
rename from Software/libremote/bsl.py
rename to Software/libremote/bsdsocket.py
diff --git a/Software/libremote/build-client.sh b/Software/libremote/build-example-client.sh
similarity index 68%
rename from Software/libremote/build-client.sh
rename to Software/libremote/build-example-client.sh
index 4ca26ea..08d964d 100755
--- a/Software/libremote/build-client.sh
+++ b/Software/libremote/build-example-client.sh
@@ -1,4 +1,4 @@
 set -e
 set -x
-python gen_proto.py libdecl-$1.json
+python gen_proto.py example
 vc +aos68k -I${NDK32}/Include_H example_client.c -o ExampleClient
diff --git a/Software/libremote/build-library.sh b/Software/libremote/build-library.sh
index c018e33..955ed28 100755
--- a/Software/libremote/build-library.sh
+++ b/Software/libremote/build-library.sh
@@ -1,4 +1,4 @@
 set -e
 set -x
-python gen_stubs.py libdecl-$1.json
+python gen_stubs.py $1
 vc +aos68k -I${NDK32}/Include_H romtag.asm library.c -nostdlib -O2 -o $1.library
diff --git a/Software/libremote/libremote.py b/Software/libremote/example.py
similarity index 100%
rename from Software/libremote/libremote.py
rename to Software/libremote/example.py
diff --git a/Software/libremote/example_client.c b/Software/libremote/example_client.c
index e8408fc..cfe139f 100644
--- a/Software/libremote/example_client.c
+++ b/Software/libremote/example_client.c
@@ -4,7 +4,7 @@
 
 #include <stdio.h>
 
-#include "proto_lib.h"
+#include "proto_example.h"
 
 struct Library *ExampleBase;
 
diff --git a/Software/libremote/gen_proto.py b/Software/libremote/gen_proto.py
index db46d04..7bb2168 100644
--- a/Software/libremote/gen_proto.py
+++ b/Software/libremote/gen_proto.py
@@ -1,7 +1,10 @@
 import json
 import sys
 
-with open(sys.argv[1], 'rt') as f:
+in_file = f'libdecl-{sys.argv[1]}.json'
+out_file = f'proto_{sys.argv[1]}.h'
+
+with open(in_file, 'rt') as f:
     libdecl = json.load(f)
 
 funcs = libdecl['funcs']
@@ -44,5 +47,5 @@ for i, func in enumerate(funcs):
 
 contents += f'#endif /* {guard} */\n'
 
-with open('proto_lib.h', 'wt') as f:
+with open(out_file, 'wt') as f:
     f.write(contents)
diff --git a/Software/libremote/gen_stubs.py b/Software/libremote/gen_stubs.py
index 3acec58..6bf4cda 100644
--- a/Software/libremote/gen_stubs.py
+++ b/Software/libremote/gen_stubs.py
@@ -1,7 +1,9 @@
 import json
 import sys
 
-with open(sys.argv[1], 'rt') as f:
+in_file = f'libdecl-{sys.argv[1]}.json'
+
+with open(in_file, 'rt') as f:
     libdecl = json.load(f)
 
 funcs = libdecl['funcs']
-- 
2.47.3

